### ISSUE TITLE
e2e flake: tests are failing on "All nodes should be ready" condition

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#21261]
I've seen a bunch of different failure, with number of tests failing due to:

     Feb 14 08:36:31.720: FAIL: All nodes should be ready after test,

The example failure is here:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/21020/kubernetes-pull-build-test-e2e-gce/28512/build-log.txt
[but there were 5+ more today in different places & suites]

**Comment by lavalamp** [comment]
I guess we need a more specific diagnosis. @dchen1107 to diagnose/delegate.

**Comment by dchen1107** [comment]
@lavalamp Yes, I checked the instance @wojtek-t posted above yesterday, and nothing is really outstanding. Meanwhile, I couldn't find other instances you mentioned since there is no link.

**Comment by wojtek-t** [comment]
@dchen1107 - what do you mean by "nothing is really outstanding". This shouldn't happen...

Regarding other failures - I can probably paste more tomorrow.

**Comment by dchen1107** [comment]
Ahh, I should make my comment more clear here. I didn't mean there is no issue.

From the log posted above, I saw after the test finished, test checked all nodes ready or not, and one node is not ready thus the test is marked not ready, and the node is not self-healing within a reasonable time, which is not acceptable. But from all available information, nothing is outstanding to me to figure out the root cause.

**Comment by dchen1107** [comment]
I managed finding one node not ready after test, and checked kubelet.log. Looks like kubelet is failed the healthcheck, and killed by supervisord:

```
2016-02-14 16:35:10,647 INFO exited: kubelet (exit status 2; expected)
2016-02-14 16:35:11,650 INFO spawned: 'kubelet' with pid 24329
2016-02-14 16:35:12,695 INFO success: kubelet entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-14 16:35:35,908 INFO exited: docker (exit status 2; expected)
2016-02-14 16:35:36,910 INFO spawned: 'docker' with pid 24371
2016-02-14 16:35:37,938 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-14 16:36:11,701 INFO exited: kubelet (exit status 2; expected)
2016-02-14 16:36:12,703 INFO spawned: 'kubelet' with pid 24473
2016-02-14 16:36:13,748 INFO success: kubelet entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

By checking kubelet log, I noticed that after last SYNCLOOP, kubelet doesn't nothing except responding /healthz, and after 5m, it is killed by supervisord:
```
I0214 16:30:09.708975    3917 kubelet.go:2371] SyncLoop (housekeeping) 
                                                                             ^^^^^^^
                                                                            Last SyncLoop ...
I0214 16:30:09.733184    3917 file.go:135] Reading config file "/etc/kubernetes/manifests/fluentd-es.yaml"
I0214 16:30:09.734095    3917 file.go:135] Reading config file "/etc/kubernetes/manifests/kube-proxy.manifest"
I0214 16:30:09.734734    3917 config.go:278] Setting pods for source file
I0214 16:30:10.018020    3917 server.go:1083] GET /healthz: (26.046Âµs) 200 [[curl/7.26.0] 127.0.0.1:44594]
I0214 16:30:20.038477    3917 server.go:1083] GET /healthz: (27.965Âµs) 200 [[curl/7.26.0] 127.0.0.1:44596]
I0214 16:30:25.974535    3917 config.go:278] Setting pods for source api
I0214 16:30:25.975143    3917 config.go:393] Receiving a new pod "netexec_e2e-tests-kubectl-rb9vg(409287ac-d338-11e5-806a-42010af00002)"
...
I0214 16:30:30.059028    3917 server.go:1083] GET /healthz: (51.272Âµs) 200 [[curl/7.26.0] 127.0.0.1:44599]
                                                                            ^^^^^^^^^
                                                                           Doing nothing, only responding /healthz
...
I0214 16:35:10.645297    3917 server.go:1083] GET /healthz: (256.193Âµs) 500
                                                                            ^^^^^^^^
                                                                           Finally killed by supervisord
goroutine 2322 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc20972b340, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc20972b340, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
net/http.Error(0x7f5cdecac9b8, 0xc20972b340, 0xc2097a5d40, 0x58, 0x1f4)
	/usr/src/go/src/net/http/server.go:1274 +0x9a
k8s.io/kubernetes/pkg/healthz.funcÂ·002(0x7f5cdecac9b8, 0xc20972b340, 0xc209438410)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/healthz/healthz.go:109 +0x46e
net/http.HandlerFunc.ServeHTTP(0xc2082816d0, 0x7f5cdecac9b8, 0xc20972b340, 0xc209438410)
	/usr/src/go/src/net/http/server.go:1265 +0x41
```

5m here makes me think kubelet might stuck at SyncPods because docker is not responding within a reasonable time, Kubelet failed healthz after 5ms. Unfortunately, supervisord omit all health failure reasons, I cannot prove my theory here. (1) I will file a separate issue to fix supervisord scripts to collect the failure reason, and collect those logs for our jenkins. 

Another found I have is that right before this, supervisord also killed docker daemon:
```
2016-02-14 16:30:31,729 INFO exited: docker (exit status 2; expected)
2016-02-14 16:30:32,731 INFO spawned: 'docker' with pid 23533
2016-02-14 16:30:33,761 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-14 16:31:32,799 INFO exited: docker (exit status 2; expected)
2016-02-14 16:31:33,801 INFO spawned: 'docker' with pid 23695
2016-02-14 16:31:34,828 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-14 16:32:53,834 INFO exited: docker (exit status 2; expected)
2016-02-14 16:32:54,836 INFO spawned: 'docker' with pid 23924
2016-02-14 16:32:55,864 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-14 16:34:14,870 INFO exited: docker (exit status 2; expected)
2016-02-14 16:34:15,873 INFO spawned: 'docker' with pid 24186
2016-02-14 16:34:16,902 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

@yujuhong recently changed docker version to docker ps so that we could reliably detect docker health, especially for https://github.com/kubernetes/kubernetes/issues/20096 But caveat is that docker ps might take long time occasionally, and supervisord might kill docker daemon unnecessarily, which might augment this issue here. (2) @yujuhong has a pending pr (https://github.com/kubernetes/kubernetes/pull/21347) to revert that change back, so that we can rule out docker healthcheck issue here. Also we need to have docker log available for debugging.

Above two issues make me think we might hit docker 1.9.X performance issue mentioned at: https://github.com/kubernetes/kubernetes/issues/16110#issuecomment-154210277, but our performance stress test couldn't reproduce it: https://github.com/kubernetes/kubernetes/issues/16110#issuecomment-180510177
We need to verify this.

(3) Also @yujuhong reminded me that recently we saw kubelet stuck at syncloop in https://github.com/kubernetes/kubernetes/issues/20418 but with docker 1.8.3 then. In this case, we might run into the hairpin mode bug. Anyway, my pending pr (https://github.com/kubernetes/kubernetes/pull/21208) should help on this case. 

cc/ @kubernetes/goog-node

**Comment by nikhiljindal** [comment]
fwiw, I am trying to reproduce https://github.com/kubernetes/kubernetes/issues/21234 and I keep running into this.
So this one seems easy to reproduce manually if you think that helps.

**Comment by gmarek** [comment]
@dchen1107 we have two NotReady nodes (as of 10 am CET),  but there were 210 NotReady ones during cluster validation phase. I disabled enormous cluster test, so you can debug.

**Comment by wojtek-t** [comment]
@dchen1107 - you can take a look into machines from:
https://pantheon.corp.google.com/compute/instances?project=kubernetes-scale

**Comment by wojtek-t** [comment]
The problematic run we are talking about is this one:
http://kubekins.dls.corp.google.com:8080/view/Scalability/job/kubernetes-e2e-gce-enormous-cluster/65/console
where only 789/1000 of nodes get ready within timeout.

**Comment by dchen1107** [comment]
@gmarek and @wojtek-t I filed a separate issue for the enormous cluster failure: https://github.com/kubernetes/kubernetes/issues/21500 Basically it is caused by crashlooping etcd-container due to oom killing.

**Comment by wojtek-t** [comment]
@dchen1107 - thanks for investigation!

**Comment by dchen1107** [comment]
@wojtek-t Do you have more incidents of NotReady nodes after the initial cluster validation? I am looking for those failures :-)

**Comment by dchen1107** [comment]
I checked nikhiljindal's cluster mentioned: https://github.com/kubernetes/kubernetes/issues/21261#issuecomment-185509681 

It is a dup issue of https://github.com/kubernetes/kubernetes/issues/20096, the kernel deadlock issue triggered by hairpin mode:

```
Message from syslogd@e2e-test-nikhiljindal-minion-7j80 at Feb 18 22:27:10 ...
 kernel:[70306.524044] unregister_netdevice: waiting for veth4f54c21 to become free. Usage count = 1
Message from syslogd@e2e-test-nikhiljindal-minion-7j80 at Feb 18 22:27:18 ...
 kernel:[70314.540048] unregister_netdevice: waiting for lo to become free. Usage count = 2
...
```

I keep this one open to make sure the situation of the node is becoming NotReady is not caused by docker 1.9.1.

**Comment by ixdy** [comment]
I think http://pr-test.k8s.io/21442/kubernetes-pull-build-test-e2e-gce/29223/ might be another example of this?

**Comment by dchen1107** [comment]
@ixdy,  I believe the failure you mentioned at https://github.com/kubernetes/kubernetes/issues/21261#issuecomment-185960665 is also hairpin mode related issue. Unfortunately there is no logs from worker nodes for that jenkins build.

**Comment by bgrant0607** [comment]
Recent e2es have failed with this symptom.

**Comment by bprashanth** [comment]
@ixdy can't tell without system logs which we don't seem to collect when build times out. 

@dchen1107 there were quite a few cases of node not ready in https://github.com/kubernetes/kubernetes/issues/20910#issuecomment-185989517, docker was crashlooping. And at least the first few weren't the hung task kernel bug from dmesg.log. I'm collecting docker.log in (https://github.com/kubernetes/kubernetes/pull/21580) because I want to know if it was some corrupt docker-local-db state type thing.

**Comment by wojtek-t** [comment]
@dchen1107 - another instance:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/21340/kubernetes-pull-build-test-e2e-gce/29611/build-log.txt

**Comment by wojtek-t** [comment]
And one more:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/21340/kubernetes-pull-build-test-e2e-gce/29704/build-log.txt

**Comment by bprashanth** [comment]
IMO this is blocked on the lack of logs when a build times out, just like https://github.com/kubernetes/kubernetes/issues/21138 (and they're probably the same root cause)

**Comment by dchen1107** [comment]
A little bit summary here:

* docker used to kill pre-running docker containers at startup stage, but recent releases, they move this logic to shutdown stage, which is good. But docker 1.9.1, since the rearchitect of docker daemon including separate docker network, persistent network &  storage state requires much long time. 
* Meanwhile, our supervisor babysitter script does respect such change in docker without waiting for docker gracefully terminated. In this case, we might run into the issue of supervisor keep shutdown the docker daemon, and immediately start it, but start failed due to previous process is not terminated. Then supervisor repeat above procedure.

I am going to send a fix shortly.
